js/compiler/compiler.go

```diff
@@ -44,103 +44,114 @@ var (
+type Compiler struct{}
 
-	return compilerInstance, err
+// New returns a new Compiler
+func New() *Compiler {
+	return &Compiler{}
 }
 
-func new() (*Compiler, error) {
-	conf := rice.Config{
-		LocateOrder: []rice.LocateMethod{rice.LocateEmbedded},
+// Transform the given code into ES5, while synchronizing to ensure only a single
+// babel instance is in use.
+func (c *Compiler) Transform(src, filename string) (code string, srcmap SourceMap, err error) {
+	var b *babel
+	if b, err = newBabel(); err != nil {
+		return
 	}
 
-	babelSrc := conf.MustFindBox("lib").MustString("babel.min.js")
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+	return b.Transform(src, filename)
+}
 
-	c := &Compiler{vm: goja.New()}
-	if _, err := c.vm.RunString(babelSrc); err != nil {
-		return nil, err
+// Compile the program in the given CompatibilityMode, optionally running pre and post code.
+func (c *Compiler) Compile(src, filename string, pre, post string,
+	strict bool, compatMode CompatibilityMode) (*goja.Program, string, error) {
+	code := pre + src + post
+	ast, err := parser.ParseFile(nil, filename, code, 0)
+	if err != nil {
+		if compatMode == CompatibilityModeExtended {
+			code, _, err = c.Transform(src, filename)
+			if err != nil {
+				return nil, code, err
+			}
+			return c.Compile(code, filename, pre, post, strict, compatMode)
+		}
+		return nil, src, err
 	}
+	pgm, err := goja.CompileAST(ast, strict)
+	return pgm, code, err
+}
 
-	c.this = c.vm.Get("Babel")
-	thisObj := c.this.ToObject(c.vm)
-	if err := c.vm.ExportTo(thisObj.Get("transform"), &c.transform); err != nil {
-		return nil, err
-	}
+type babel struct {
+	vm        *goja.Runtime
+	this      goja.Value
+	transform goja.Callable
+	mutex     sync.Mutex //TODO: cache goja.CompileAST() in an init() function?
+}
+
+func newBabel() (*babel, error) {
+	var err error
 
-	return c, nil
+	once.Do(func() {
+		conf := rice.Config{
+			LocateOrder: []rice.LocateMethod{rice.LocateEmbedded},
+		}
+		babelSrc := conf.MustFindBox("lib").MustString("babel.min.js")
+		vm := goja.New()
+		if _, err = vm.RunString(babelSrc); err != nil {
+			return
+		}
+
+		this := vm.Get("Babel")
+		bObj := this.ToObject(vm)
+		babl = &babel{vm: vm, this: this}
+		if err = vm.ExportTo(bObj.Get("transform"), &babl.transform); err != nil {
+			return
+		}
+	})
+
+	return babl, err
 }
 
-// Transform the given code into ES5.
-func (c *Compiler) Transform(src, filename string) (code string, srcmap SourceMap, err error) {
-	opts := make(map[string]interface{})
-	for k, v := range DefaultOpts {
-		opts[k] = v
-	}
+func (b *babel) Transform(src, filename string) (code string, srcmap SourceMap, err error) {
+	opts := DefaultOpts
 	opts["filename"] = filename
 
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-
 	startTime := time.Now()
-	v, err := c.transform(c.this, c.vm.ToValue(src), c.vm.ToValue(opts))
+	v, err := b.transform(b.this, b.vm.ToValue(src), b.vm.ToValue(opts))
 	if err != nil {
-		return code, srcmap, err
+		return
```

Comment:
NO, I really [naked returns](https://www.ardanlabs.com/blog/2013/10/functions-and-naked-returns-in-go.html) and especially in not short functions and in combination with the way `:=` works it's basically creating the perfect place for strange and hard to reason behaviour

Reply:
A few points:
- No need to shout.
- You missed a verb after "really" there, but I'm figuring it's not positive. :)
- I agree with you that naked returns can be difficult to track in long functions, but `babel.Transform` fits comfortably on a single page so I would argue it's quite easy to follow, even though it's more medium-sized.
- The shadowing on line 139 wasn't intentional and should be fixed, even though it has no effect in this case.
- The alternative to using naked returns in this function becomes very verbose and repetitive because of its three return values and 5 `return`s. The original implementation used them as well, though not for errors.
- Naked returns seem like a fine language feature when the scope is limited, as in this case. It's essentially just a convenience method of declaring return variables upfront and cutting down on repetition for each return, which kind of makes sense in this case.

I'm wondering what the other guys think, but let me know if you still want this fixed.

Reply:
* sorry, don't remember if I meant it to be in caps or have I just not released the shift in time
* `dislike`
* You are *not* shadowing the `err` because naked returns and `:=` are *magic*. If you would've shadowed one of the "return vars" and used a naked return there would've been a *compile time* error - good design on top of the (IMO) bad one. 
* The above or the fact that the previous code `return`s were even more unreadable doesn't change the fact that I think that  the returns should be  of the kind `return nil, nil, err` or `return code, srcmap, nil` (the only two valid cases IMO). While I do like the naming of the return types for documentation sake I would argue it's not a good practice to use naked returns .. ever .. I at least  see no point beside making it confusing or in rare cases where the returns are actually ... long 
* Separate point: ATM (or for that matter forever, I think) k6 doesn't do anything with the sourcemaps, #1107  mentions it. While I think that this will change in the future ... it's unlikely (IMO) to be soon and it probably will be behind a flag as it does have a performance penalty. Given that the flag will probably be off by default we might just not populate it now and than when we add the flag to add the code back ? cc @na--

Reply:
> If you would've shadowed one of the "return vars" and used a naked return there would've been a compile time error

Hhmm right, I wondered why it didn't error. It's good practice to not do it anyway, though.

I still think this function is more readable with naked returns, and the potential risks of misusing it are minimal, but I'll follow your advice and remove them.

Removing sourcemaps makes sense if they're not used, but I'd rather leave that for another PR.

Reply:
Fixed in 19617a47.